Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-327.28.3.el7.x86_64)
 GCC:gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Sha: 33e7f34bf10443aadb66946cc16e0007848f5de6
Number of Error(s): 9
../../RelWithDebInfo/libs/libgraph_lcr.so: undefined reference to `setProcessAborted(unsigned int)
collect2: error: ld returned 1 exit status
make[2]: *** [thorlcr/master/CMakeFiles/thormaster_lcr.dir/build.make:167: RelWithDebInfo/bin/thormaster_lcr] Error 1
make[1]: *** [CMakeFiles/Makefile2:17135: thorlcr/master/CMakeFiles/thormaster_lcr.dir/all] Error 2
../../RelWithDebInfo/libs/libgraph_lcr.so: undefined reference to `setProcessAborted(unsigned int)
collect2: error: ld returned 1 exit status
make[2]: *** [thorlcr/slave/CMakeFiles/thorslave_lcr.dir/build.make:164: RelWithDebInfo/bin/thorslave_lcr] Error 1
make[1]: *** [CMakeFiles/Makefile2:17368: thorlcr/slave/CMakeFiles/thorslave_lcr.dir/all] Error 2
make: *** [Makefile:163: all] Error 2

[Build error:](https://gist.github.com/c833053757e29561658acaad69bb687a)
Build log [Build log](https://gist.github.com/HPCCSmoketest/c833053757e29561658acaad69bb687a#file-result-19-11-25-08-28-23-log)

--- a/system/jlib/jexcept.cpp
+++ b/system/jlib/jexcept.cpp
@@ -1101,6 +1101,12 @@ static void throwSigSegV()
 }
 #endif
 
+static std::atomic<bool> processAborted { false };
+void jlib_decl setProcessAborted(bool _abortVal)
+{
+    processAborted = _abortVal;
+}
+
 NO_SANITIZE("alignment") void excsighandler(int signum, siginfo_t *info, void *extra)
 {
     static byte nested=0;
@@ -1354,7 +1360,10 @@ NO_SANITIZE("alignment") void excsighandler(int signum, siginfo_t *info, void *e
         if ( SEHHandler->fireException(new CSEHException(signum,s.str())) )
             return;
     }
-    raise(signum);
+    if ( (SIGABRT == signum) && (processAborted) )
+        _exit(128 + SIGABRT);
+    else
+        raise(signum);
 #endif
     nested--;
 }

--- a/system/jlib/jexcept.hpp
+++ b/system/jlib/jexcept.hpp
@@ -129,6 +129,7 @@ void jlib_decl *setSEHtoExceptionHandler(IExceptionHandler *handler); // sets ha
 void jlib_decl setTerminateOnSEHInSystemDLLs(bool set=true);
 void jlib_decl setTerminateOnSEH(bool set=true);
 
+void jlib_decl setProcessAborted(bool _abortVal);
 
 __declspec(noreturn) void jlib_decl throwUnexpectedException(const char * file, unsigned line) __attribute__((noreturn));
 __declspec(noreturn) void jlib_decl throwUnexpectedException(const char * where, const char * file, unsigned line) __attribute__((noreturn));

--- a/thorlcr/graph/thgraph.cpp
+++ b/thorlcr/graph/thgraph.cpp
@@ -1727,6 +1727,13 @@ void CGraphBase::abort(IException *e)
 
         abortException.set(e);
         aborted = true;
+
+        if (!job.getOptBool("coreOnAbort", false))
+        {
+            // prevent dtors throwing ... __verbose_terminate_handler()->abort()->raise()->core files
+            setProcessAborted(true);
+        }
+
         graphCancelHandler.cancel(0);
 
         if (0 == containers.count())

--- a/thorlcr/master/thgraphmanager.cpp
+++ b/thorlcr/master/thgraphmanager.cpp
@@ -696,6 +696,9 @@ void CJobManager::run()
         { 
             reply(workunit, wuid, MakeStringException(0, "Unknown exception"), agentep, false); 
         }
+
+        // reset for next job
+        setProcessAborted(false);
     }
     delete dp;
     jobq.clear();

--- a/thorlcr/msort/tsorts.cpp
+++ b/thorlcr/msort/tsorts.cpp
@@ -798,12 +798,21 @@ public:
     }
     ~CThorSorter()
     {
-        stop();
-        ActPrintLog(activity, "Joining Sort Slave Server");
-        verifyex(threaded.join(10*60*1000));
-        myendpoint.set(NULL,0);
-        rowArray.kill();
-        ActPrintLog(activity, "~CThorSorter");
+        try
+        {
+            stop();
+            ActPrintLog(activity, "Joining Sort Slave Server");
+            verifyex(threaded.join(10*60*1000));
+            myendpoint.set(NULL,0);
+            rowArray.kill();
+            ActPrintLog(activity, "~CThorSorter");
+        }
+        catch (IException *e)
+        {
+            e->Release();
+        }
+        // do we ignore other exceptions (...) here ?
+        // if so, we may have std::terminate call abort() ...
     }
 // ISortSlaveMP
     virtual bool Connect(unsigned _partno, unsigned _numnodes)

--- a/thorlcr/slave/slavmain.cpp
+++ b/thorlcr/slave/slavmain.cpp
@@ -1645,10 +1645,19 @@ public:
     }
     ~CJobListener()
     {
-        for (unsigned sc=1; sc<channelsPerSlave; sc++)
-            mpServers.item(sc).stop();
-        mpServers.kill();
-        stop();
+        try
+        {
+            for (unsigned sc=1; sc<channelsPerSlave; sc++)
+                mpServers.item(sc).stop();
+            mpServers.kill();
+            stop();
+        }
+        catch (IException *e)
+        {
+            e->Release();
+        }
+        // do we ignore other exceptions (...) here ?
+        // if so, we may have std::terminate call abort() ...
     }
     void stop()
     {
@@ -1867,6 +1876,10 @@ public:
                         }
                         jobs.removeExact(job);
                         PROGLOG("QueryDone, removed %s from jobs", key.get());
+
+                        // reset for next job
+                        setProcessAborted(false);
+
                         if (exception)
                             throw exception.getClear(); // NB: this will cause exception to be part of the reply to master
 

